stan/math/rev/fun/accumulator.hpp

```diff
@@ -30,7 +24,7 @@ namespace math {
 template <>
 class accumulator<var> {
  private:
-  std::vector<var> buf_;
+  std::vector<var, arena_allocator<var>> buf_;
```

Comment:
That looks like a nice trick. However note that pushing things into arena vector wastes memory, as reallocations do not clean up previously used memory. So the memory is essentially leaked until the end of iteration when all the arena memory is recovered. 

Is this what gave you this additional performance bump? If so, I think it is fine to leave in despite it potentially using double the amount of memory it needs.

Reply:
So removing it removes the speedup, but I'm actually not sure how large this thing can get (it get's called in loops really often in Stan programs) so until I get a better feel for that I'm going to put this back to a regular vector. Running the tests we miss the speedup but things stay the same as they are now.

### This PR

```
-------------------------------------------------------------------------
Benchmark                               Time             CPU   Iterations
-------------------------------------------------------------------------
acc_var_bench/2/manual_time          76.6 ns          140 ns      8619322
acc_var_bench/4/manual_time           108 ns          178 ns      6588146
acc_var_bench/8/manual_time           139 ns          221 ns      4854991
acc_var_bench/16/manual_time          184 ns          298 ns      3832329
acc_var_bench/32/manual_time          255 ns          411 ns      2728349
acc_var_bench/64/manual_time          402 ns          663 ns      1743512
acc_var_bench/128/manual_time         671 ns         1224 ns       999114
acc_var_bench/256/manual_time        1143 ns         2190 ns       615175
acc_var_bench/512/manual_time        2053 ns         4092 ns       340354
acc_var_bench/1024/manual_time       3814 ns         7776 ns       182441
acc_var_bench/2048/manual_time       7296 ns        15141 ns        95369
acc_var_bench/4096/manual_time      14167 ns        29812 ns        52437
```

### Develop

```
-------------------------------------------------------------------------
Benchmark                               Time             CPU   Iterations
-------------------------------------------------------------------------
acc_var_bench/2/manual_time          72.2 ns          136 ns     10078474
acc_var_bench/4/manual_time           102 ns          171 ns      6989435
acc_var_bench/8/manual_time           124 ns          202 ns      5566992
acc_var_bench/16/manual_time          167 ns          268 ns      4120880
acc_var_bench/32/manual_time          235 ns          396 ns      2889932
acc_var_bench/64/manual_time          398 ns          677 ns      1803766
acc_var_bench/128/manual_time         660 ns         1203 ns      1078776
acc_var_bench/256/manual_time        1129 ns         2170 ns       621440
acc_var_bench/512/manual_time        2046 ns         4051 ns       339553
acc_var_bench/1024/manual_time       3832 ns         7747 ns       182735
acc_var_bench/2048/manual_time       7307 ns        15026 ns        95569
acc_var_bench/4096/manual_time      14114 ns        29520 ns        48856
```